File: go/message_decryptor.go
package webexmessagehandler

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/go-jose/go-jose/v4"
)

// MessageDecryptor decrypts encrypted Webex message activities using KMS keys.
type MessageDecryptor struct {
	kmsClient *KmsClient
	logger    Logger
}

// NewMessageDecryptor creates a new MessageDecryptor.
func NewMessageDecryptor(kmsClient *KmsClient, logger Logger) *MessageDecryptor {
	if logger == nil {
		logger = NoopLogger()
	}
	return &MessageDecryptor{kmsClient: kmsClient, logger: logger}
}

// DecryptActivity decrypts an encrypted Mercury activity.
// Returns a copy with decrypted DisplayName and Content fields.
// If the activity is not encrypted (no EncryptionKeyURL), returns as-is.
func (md *MessageDecryptor) DecryptActivity(ctx context.Context, activity MercuryActivity) (MercuryActivity, error) {
	// Locate encryption key URL
	encryptionKeyURL := activity.EncryptionKeyURL
	if encryptionKeyURL == "" {
		encryptionKeyURL = activity.Object.EncryptionKeyURL
	}
	if encryptionKeyURL == "" {
		encryptionKeyURL = activity.Target.EncryptionKeyURL
	}

	// Not encrypted
	if encryptionKeyURL == "" {
		return activity, nil
	}

	// Fetch the key from KMS
	key, err := md.kmsClient.GetKey(ctx, encryptionKeyURL)
	if err != nil {
		return activity, NewDecryptionErrorWithCause(
			fmt.Sprintf("Failed to fetch encryption key from %s", encryptionKeyURL), err,
		)
	}

	// Copy activity (value type, so already a copy)
	decrypted := activity

	// Decrypt displayName
	if decrypted.Object.DisplayName != "" {
		plaintext, err := decryptJWE(decrypted.Object.DisplayName, key)
		if err != nil {
			md.logger.Warn(fmt.Sprintf("Failed to decrypt displayName in activity %s: %v", activity.ID, err))
		} else {
			decrypted.Object.DisplayName = string(plaintext)
		}
	}

	// Decrypt content
	if decrypted.Object.Content != "" {
		plaintext, err := decryptJWE(decrypted.Object.Content, key)
		if err != nil {
			md.logger.Warn(fmt.Sprintf("Failed to decrypt content in activity %s: %v", activity.ID, err))
		} else {
			decrypted.Object.Content = string(plaintext)
		}
	}

	// Decrypt card-action inputs. On cardAction/submit activities object.inputs
	// is a JWE string encrypted under the same key; the decrypted plaintext is a
	// JSON object of the card's form values.
	if decrypted.Object.InputsEncrypted != "" {
		plaintext, err := decryptJWE(decrypted.Object.InputsEncrypted, key)
		if err != nil {
			md.logger.Warn(fmt.Sprintf("Failed to decrypt inputs in activity %s: %v", activity.ID, err))
		} else {
			var inputs map[string]interface{}
			if err := json.Unmarshal(plaintext, &inputs); err != nil {
				md.logger.Warn(fmt.Sprintf("Failed to parse decrypted inputs in activity %s: %v", activity.ID, err))
			} else {
				decrypted.Object.Inputs = inputs
			}
		}
	}

	return decrypted, nil
}

func decryptJWE(token string, key *jose.JSONWebKey) ([]byte, error) {
	// Webex messages may use "dir" (direct CEK) or "A256KW" (key wrapping)
	jweObj, err := jose.ParseEncrypted(token, []jose.KeyAlgorithm{jose.DIRECT, jose.A256KW}, []jose.ContentEncryption{jose.A256GCM})
	if err != nil {
		return nil, fmt.Errorf("failed to parse JWE: %w", err)
	}
	plaintext, err := jweObj.Decrypt(key.Key)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt JWE: %w", err)
	}
	return plaintext, nil
}

File: node/src/message-decryptor.ts
import * as jose from 'node-jose';
import { KmsClient } from './kms-client.js';
import { Logger, noopLogger } from './logger.js';
import { MercuryActivity } from './types.js';
import { DecryptionError } from './errors.js';

export class MessageDecryptor {
  private kmsClient: KmsClient;
  private logger: Logger;

  constructor({ kmsClient, logger }: { kmsClient: KmsClient; logger?: Logger }) {
    this.kmsClient = kmsClient;
    this.logger = logger || noopLogger;
  }

  async decryptActivity(activity: MercuryActivity): Promise<MercuryActivity> {
    try {
      // Get encryptionKeyUrl from one of three possible locations
      const encryptionKeyUrl =
        activity.encryptionKeyUrl ||
        activity.object.encryptionKeyUrl ||
        activity.target.encryptionKeyUrl;

      // If no key URL, activity is not encrypted
      if (!encryptionKeyUrl) {
        return activity;
      }

      // Fetch the key from KMS
      let key: jose.JWK.Key;
      try {
        key = await this.kmsClient.getKey(encryptionKeyUrl);
      } catch (error) {
        throw new DecryptionError(
          `Failed to fetch encryption key from ${encryptionKeyUrl}: ${error instanceof Error ? error.message : String(error)}`
        );
      }

      // Create a shallow copy of the activity with decrypted object
      const decryptedActivity: MercuryActivity = {
        ...activity,
        object: { ...activity.object },
      };

      // Decrypt displayName if it exists and is a non-empty string
      if (decryptedActivity.object.displayName && typeof decryptedActivity.object.displayName === 'string' && decryptedActivity.object.displayName.length > 0) {
        try {
          const result = await jose.JWE.createDecrypt(key).decrypt(decryptedActivity.object.displayName);
          decryptedActivity.object.displayName = result.payload.toString('utf8');
        } catch (error) {
          this.logger.warn(
            `Failed to decrypt displayName in activity ${activity.id}: ${error instanceof Error ? error.message : String(error)}`
          );
        }
      }

      // Decrypt content if it exists and is a non-empty string
      if (decryptedActivity.object.content && typeof decryptedActivity.object.content === 'string' && decryptedActivity.object.content.length > 0) {
        try {
          const result = await jose.JWE.createDecrypt(key).decrypt(decryptedActivity.object.content);
          decryptedActivity.object.content = result.payload.toString('utf8');
        } catch (error) {
          this.logger.warn(
            `Failed to decrypt content in activity ${activity.id}: ${error instanceof Error ? error.message : String(error)}`
          );
        }
      }

      // Decrypt inputs if it exists and is a non-empty string (JWE-encrypted card action inputs)
      if (decryptedActivity.object.inputs && typeof decryptedActivity.object.inputs === 'string' && decryptedActivity.object.inputs.length > 0) {
        try {
          const result = await jose.JWE.createDecrypt(key).decrypt(decryptedActivity.object.inputs);
          const plaintextInputs = result.payload.toString('utf8');
          decryptedActivity.object.inputs = JSON.parse(plaintextInputs);
        } catch (error) {
          this.logger.warn(
            `Failed to decrypt inputs in activity ${activity.id}: ${error instanceof Error ? error.message : String(error)}`
          );
          // On decrypt failure, leave inputs as encrypted string (will be handled as empty in handler)
        }
      }

      return decryptedActivity;
    } catch (error) {
      // Re-throw DecryptionError as-is, wrap other errors
      if (error instanceof DecryptionError) {
        throw error;
      }
      throw new DecryptionError(
        `Error decrypting activity ${activity.id}: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }
}
